### PR TITLE
Addresses #4538, wrap phase change material properties

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

Tests demonstrating https://github.com/NREL/OpenStudio/pull/4610

Link to relevant GitHub Issue(s) if appropriate:

Link to the **Linux.deb** installer to use for CI Testing. If not set, it will default to latest official release.
[OpenStudio Installer]: http://openstudio-ci-builds.s3-website-us-west-2.amazonaws.com/PR-4610/OpenStudio-3.4.1-alpha%2Be9677bc77b-Ubuntu-18.04.deb

This Pull Request is concerning:

 - [x] **Case 1 - `NewTest`:** a new test for a new model API class,

----------------------------------------------------------------------------------------------------------

### Case 1: New test for a new model API class

Please include which class(es) you are adding a test to specifically test for.
Include a link to the OpenStudio Pull Request in which you are adding the new classes, or the class itself if already on develop.

#### Work Checklist

The following has been checked to ensure compliance with the guidelines:

 - [x] Tests pass either:
     - [ ] with official OpenStudio release (include version):
         - [ ] A matching OSM test has been added from the successful run of the Ruby one with the official OpenStudio release
         - [ ] The label `AddedOSM` has been added to this PR
         - [ ] All new `out.osw` have been committed

     - [x] with current develop (incude SHA):
         - [x] The label `PendingOSM` has been added to this PR
         - [x] A matching OSM test has not yet been added because the official release is pending, but `model_tests.rb` has a TODO.
            ```ruby
            def test_airterminal_cooledbeam_rb
              result = sim_test('airterminal_cooledbeam.rb')
            end

            # TODO: To be added in the next official release after: 2.5.0
            # def test_airterminal_fourpipebeam_osm
            #   result = sim_test('airterminal_fourpipebeam.osm')
            # end
            ```
        - [x] No `out.osw` have been committed as they need to be run with an official OpenStudio version

 - [x] **Ruby test is stable**: when run multiple times on the same machine, it produces the same total site kBTU.
    Please paste the heatmap png generated after running the following commands:
     - [x] I ensured that I assign systems/loads/etc in a repeatable manner (eg: if I assign stuff to thermalZones, I do `model.getThermalZones.sort_by{|z| z.name.to_s}.each do ...` so I am sure I put the same ZoneHVAC systems to the same zones regardless of their order)
     - [x] I tested stability using `process_results.py` (see `python process_results.py --help` for usage).
    Please paste the text output or heatmap png generated after running the following commands:
        ```bash
        # Clean up all custom-tagged OSWs
        python process_results.py test-stability clean
        # Run your test 5 times in a row. Replace `testname_rb` (eg `airterminal_fourpipebeam_rb`)
        python process_results.py test-stability run -n testname_rb
        # Check that they all passed
        python process_results.py test-status --tagged
        # Check site kBTU differences
        python process_results.py heatmap --tagged

        ```

 - [x] **Object has been added to `autosize_hvac.rb` to ensure the autosizedXXX values methods do work**

----------------------------------------------------------------------------------------------------------

### Review Checklist

 - [x] Code style (indentation, variable names, strip trailing spaces)
 - [x] Functional code review (it has to work!)
 - [x] Matching OSM test has been added or `# TODO` added to `model_tests.rb`
 - [x] Appropriate `out.osw` have been committed
 - [x] Test is stable
 - [x] Object is tested in `autosize_hvac` as appropriate
 - [x] The appropriate labels have been added to this PR:
   - [x] One of: `NewTest`, `TestFix`, `NewTestForExisting`, `Other`
   - [x] If `NewTest`: add `PendingOSM` or `AddedOSM`

### CODE REVIEW

**Comment by jmarrec** [comment]
I updated the installer link to the latest from https://github.com/NREL/OpenStudio/pull/4610 and rebooted the Test PR Changes run

File: model/simulationtests/phase_change.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+heat_balance_algorithm = model.getHeatBalanceAlgorithm
+heat_balance_algorithm.setAlgorithm('ConductionFiniteDifference')
+
+# make a 1 story, 100m X 50m, 5 zone core/perimeter building
+model.add_geometry({ 'length' => 100,
+                     'width' => 50,
+                     'num_floors' => 1,
+                     'floor_to_floor_height' => 4,
+                     'plenum_height' => 1,
+                     'perimeter_zone_depth' => 3 })
```
Comment (jmarrec) [inline model/simulationtests/phase_change.rb]: Make this smaller please. One zone should be enough no?

File: model/simulationtests/phase_change.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+# add windows at a 40% window-to-wall ratio
+model.add_windows({ 'wwr' => 0.4,
+                    'offset' => 1,
+                    'application_type' => 'Above Floor' })
+
+# add ASHRAE System type 01, PTAC, Residential
+model.add_hvac({ 'ashrae_sys_num' => '01' })
```
Comment (jmarrec) [inline model/simulationtests/phase_change.rb]: maybe ideal air loads would be faster. see adiabatic_construction.rb for an example.

File: model/simulationtests/phase_change.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+# assign material property phase change properties
+model.getSurfaces.each do |surface|
+  if surface.surfaceType.downcase == 'wall'
+
+    surface.construction.get.to_Construction.get.layers.each do |layer|
+      next unless layer.to_StandardOpaqueMaterial.is_initialized
+
+      opt_phase_change = layer.createMaterialPropertyPhaseChange
+      if opt_phase_change.is_initialized
```
Comment (jmarrec) [inline model/simulationtests/phase_change.rb:66]: change to raise if !initialized

File: model/simulationtests/phase_change.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      end
+    end
+
+  elsif surface.surfaceType.downcase == 'floor'
+
+    surface.construction.get.to_Construction.get.layers.each do |layer|
+      next unless layer.to_StandardOpaqueMaterial.is_initialized
+
+      opt_phase_change_hysteresis = layer.createMaterialPropertyPhaseChangeHysteresis
+      if opt_phase_change_hysteresis.is_initialized
```
Comment (jmarrec) [inline model/simulationtests/phase_change.rb:85]: change to raise if !initialized

**Review by jmarrec — CHANGES_REQUESTED** [review]
Very minimal changes needed, otherwise LGTM

File: model/simulationtests/phase_change.rb (hunk 3 above)
Comment (joseph-robertson) [reply model/simulationtests/phase_change.rb:66]: But doesn't "createXXX" return false if a material property was already created for a given material?

Comment (jmarrec) [reply model/simulationtests/phase_change.rb:66]: I suppose if you have materials that are shared between constructions that is true yes.

**Pull request merged**